### ISSUE TITLE
Visualizing data

### ISSUE THREAD

**Issue by glisca** [eth-ait/dip18#4]
Hi there,

I am trying to visualize the IMU data contained in DIPIMUandOthers [1] and I suspect that the current version of the _live_demo/inference_server.py_ [2] and the format of the data are incompatible. The _*.pkl_ files contain a dictionary which the code tries to parse into an _ndarray_.
Could you please comment on this?
Many thanks in advance!

[1] https://psfiles.is.tuebingen.mpg.de/downloads/dip/DIPIMUandOthers_zip
[2] https://github.com/eth-ait/dip18/blob/master/live_demo/inference_server.py

**Comment by kaufManu** [comment]
Hi glisca,

It's possible that the underlying data representation changed - I'll check and get back to you. In the meantime it should be straight-forward to work around this - just extract the ground-truth poses from the incoming data.

**Comment by glisca** [comment]
Hi @kaufManu,

and thanks for your fast reply and indication! Yes, parsing parsing only the ground truth _'gt'_ works well.

**Comment by YinghaoHuang91** [comment]
Hi @glisca,

I just checked the data file again. For DIP_IMU.zip, every pickle file contains a dictionary, with fields [sop, sip, gt, imu]. You can switch among them to show the results obtained via different method. Can you please be more specific about what's wrong with it? 

Thanks!

Yinghao

**Comment by glisca** [comment]
Hi @YinghaoHuang91 ,

thanks a lot for taking the time and having a look at the data! I followed the instruction from the [Visualization](https://github.com/eth-ait/dip18/blob/master/live_demo/README.md#visualization) section and:
- added the _vizualize_data.cs_ to the SMPL component (after I created it in Unity3D).
- entered the system path of the _01.pkl_ file (_/home/noname/data/DIP_IMU/s_01/01.pkl_ for example) into **_OURS_** field (of the _Vizualize_data (Script)_ of my SMPL component from Unity3D)
- run the scene in Unity3D with Play.
- noticed that the server printed these messages:
  `tuple index out of range`
  `client terminated connection`
  I assume that this happens because in this [line](https://github.com/eth-ait/dip18/blob/4f4a9a23a3ffd9ae15f834783bb78053cd532ba6/live_demo/inference_server.py#L277) the dictionary loaded from the _01.pkl_ file is parsed into an narray with shape `()` and hence this [line](https://github.com/eth-ait/dip18/blob/4f4a9a23a3ffd9ae15f834783bb78053cd532ba6/live_demo/inference_server.py#L295) fails to extract the first element of the shape (which is in fact `()`).

When taking only one entry of the dictionary `sop` or `gt` or `sip` (by replacing the [line](https://github.com/eth-ait/dip18/blob/4f4a9a23a3ffd9ae15f834783bb78053cd532ba6/live_demo/inference_server.py#L277) with `poses = np.array(sample_sequence['gt'])` for example) then the visualization works just fine.

When taking the `imu` entry of the dictionary (by replacing the [line](https://github.com/eth-ait/dip18/blob/4f4a9a23a3ffd9ae15f834783bb78053cd532ba6/live_demo/inference_server.py#L277) with `poses = np.array(sample_sequence['imu'])`) the Unity3D client trows `JsonReaderException: Unexpected character encountered while parsing value: [. Path 'pose', line 1, position 10.` from this [line](https://github.com/eth-ait/dip18/blob/4f4a9a23a3ffd9ae15f834783bb78053cd532ba6/live_demo/visualize_data.cs#L81). Further I noticed that the narray [poses](https://github.com/eth-ait/dip18/blob/4f4a9a23a3ffd9ae15f834783bb78053cd532ba6/live_demo/inference_server.py#L277) (resulted from `imu` entry of the dictionary) contains  `nan` values.

Any suggestions are more than welcome :) Thanks again!

Gheorghe

**Comment by YinghaoHuang91** [comment]
Hi @glisca,

Actually different from other ones being the SMPL pose parameters, the "imu" field stores the *original* IMU readings of DIP_IMU dataset. That's why there are some "nan" values. I think you firstly need to run the trained model on DIP_IMU before you can visualize it.

Yinghao

**Comment by glisca** [comment]
Hi @YinghaoHuang91 

ok, I'll run the trained model.

Still related to data visualization, is there a way to visualize the training data? ... for example the _/synthetic60FPS/AMASS_ACCAD//s011walkdog_dynamics.pkl_ file?

Gheorghe

**Comment by YinghaoHuang91** [comment]
Hi @glisca,

The training data is always composed of pairs of synthetic IMU readings and corresponding (partial, major skeleton) SMPL pose parameters. You can simply visualize the SMPL pose part of the training data just like how you visualize the DIP result. Or you can follow the way shown in other project like HMR: https://github.com/akanazawa/hmr.

Yinghao

**Comment by glisca** [comment]
Hi @YinghaoHuang91 ,

I attempt to visualize the `poses` stored into `synthetic60FPS/AMASS_CMU_Kitchen/Subject1brownies__dynamics.pkl` synthetic IMU data file without success, so far. Meanwhile I noticed that the shape of the ndarray parsed from the `poses` dictionary's entry is `(10572, 135)`. I guess that `10572` is the number of poses and an SMPL pose is represented trough `135` values.
From this code [`_poseUpdater.setNewPose(p.pose.ToArray());`](https://github.com/eth-ait/dip18/blob/4f4a9a23a3ffd9ae15f834783bb78053cd532ba6/live_demo/visualize_data.cs#L81) and this [comment](https://github.com/eth-ait/dip18/blob/4f4a9a23a3ffd9ae15f834783bb78053cd532ba6/live_demo/SMPLPoseUpdater.cs#L327) I understand that the [SMPLPoseUpdater](https://github.com/eth-ait/dip18/blob/4f4a9a23a3ffd9ae15f834783bb78053cd532ba6/live_demo/SMPLPoseUpdater.cs#L166) expects an SMPL pose represented through `72` values rather than `135`.
1. Could you please point me to the meaning of the `135` values representing an SMPL pose? It feels that the answer to this question should be trivial, and so far I didn't find any hint towards it.
2. How can I convert the `135` values SMPL pose representation into a `72` values SMPL pose represenation?

Gheorghe

**Comment by YinghaoHuang91** [comment]
Hi @glisca,

SMPL has 24 joints altogether. Expressed in axis-angle format (3-d) that's 24x3=72-d for whole-body pose configuration. In our project, we only consider the 15 major joints of the skeleton, discarding the extra 24-15=9 joints on hands and feet. Expressed in rotation-matrix format (9-d) that's 15x9=135-d. 

You can firstly convert the pose back from rotation matrix to axis-angle, then fill in zero for the missing joints. I believe that's how we are currently doing. You can find it in the visualization code somewhere, I think. I hope you question gets resolved. 

Best!

Yinghao

**Comment by glisca** [comment]
Hi @YinghaoHuang91 

thanks a lot for clarification. I followed your indications and I am able to visualize the `pose` entries contained in the `synthetic60FPS` dataset. 
I chose these 15 joints [1, 2, 3, 4, 5, 6, 7, 8, 10, 11, 15, 16, 17, 18, 19] as the 15 major joints of the skeleton and I am not sure that I choose them correctly. Could you please check them and confirm or correct them?

Thanks again!
Gheorghe

**Comment by YinghaoHuang91** [comment]
Hi @glisca,

Sorry for the late response. You can check the major joints in this part of the code:
![image](https://user-images.githubusercontent.com/13058193/68614501-e0f41a00-0475-11ea-8e80-65704483ad18.png)

Hope it helps.

Yinghao

**Comment by glisca** [comment]
Thanks a lot dear @YinghaoHuang91 ! It's way clearer now :)

**Comment by kaufManu** [comment]
Closing, feel free to reopen for follow-up questions.
